### PR TITLE
Fix the bug in `F1AdaptiveThreshold` which occurs only when there are no anomalous images in a validation set

### LINKED ISSUE #2433: [Bug]: F1AdaptiveThreshold incorrectly selects minimum value when no anomalous images are in validation set
### Describe the bug

The `F1AdaptiveThreshold` is designed to select the optimal threshold for separating normal and anomalous classes based on the F1 score. Ideally, this optimal threshold should correspond to the "maximum" predicted value when all images are normal. However, the current implementation incorrectly uses the "minimum" predicted value instead. This error leads to an abnormally low threshold when the model is trained exclusively with normal images, resulting in an excessively high false alarm rate.

### Dataset

N/A

### Model

N/A

### Steps to reproduce the behavior

```
labels = torch.tensor([0, 0, 0, 0])
preds = torch.tensor([1.0, 2.0, 3.0, 4.0])

adaptive_threshold = F1AdaptiveThreshold(default_value=0.5)
threshold = adaptive_threshold(preds, labels)

# now threshold is tensor(1.0) but it should be tensor(4.0) because all images are normal
```

### OS information

OS information:
- OS: macOS Sequoia 15.1.1
- Python version: 3.10.14
- Anomalib version: 2.0.0dev
- PyTorch version: 2.5.1
- CUDA/cuDNN version: N/A
- GPU models and configuration: N/A
- Any other relevant information: N/A

### Expected behavior

When all images are normal, the optimized value of `F1AdaptiveThreshold` should be the "maximum" predicted value.

In other words, the `F1AdaptiveThreshold` should pass the following test case:
```
labels = torch.tensor([0, 0, 0, 0])
preds = torch.tensor([1.0, 2.0, 3.0, 4.0])

adaptive_threshold = F1AdaptiveThreshold(default_value=0.5)
adaptive_threshold.update(preds, labels)
threshold_value = adaptive_threshold.compute()

assert threshold_value == 4.0
```

### Screenshots

_No response_

### Pip/GitHub

GitHub

### What version/branch did you use?

https://github.com/openvinotoolkit/anomalib/commit/bcc0b439f616b13a8629cb64d8bf0f88fc9083a8

### Configuration YAML

```yaml
N/A
```

### Logs

```shell
N/A
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tanemaki**
## 📝 Description

- Fix the bug in `F1AdaptiveThreshold` which occurs only when there are no anomalous images in a validation set
- 🛠️ Fixes #2433 

## ✨ Changes

Select what type of change your PR is:

- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [x] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: tests/unit/metrics/test_adaptive_threshold.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
     [
         (torch.Tensor([0, 0, 0, 1, 1]), torch.Tensor([2.3, 1.6, 2.6, 7.9, 3.3]), 3.3),  # standard case
         (torch.Tensor([1, 0, 0, 0]), torch.Tensor([4, 3, 2, 1]), 4),  # 100% recall for all thresholds
+        (torch.Tensor([1, 1, 1, 1]), torch.Tensor([4, 3, 2, 1]), 1),  # use minimum value when all images are anomalous
```
Comment (djdameln) [inline tests/unit/metrics/test_adaptive_threshold.py:21]: Maybe we could add an explicit check in the metric implementation for this case, similar to the check you added for only normal images.

**Review by djdameln — CHANGES_REQUESTED** [review]
Thanks for your PR. This is a useful addition. Previously the metric would implicitly return the maximum value when all images were normal, but as it shows this was not a robust approach. Explicitly checking these edge cases is much better. I do have 2 suggestions for improvement, see below.

File: src/anomalib/metrics/threshold/f1_adaptive_threshold.py [hunk 2]
```diff
@@ -75,6 +75,11 @@ def compute(self) -> torch.Tensor:
             # special case where recall is 1.0 even for the highest threshold.
             # In this case 'thresholds' will be scalar.
             self.value = thresholds
+        elif not any(1 in batch for batch in self.target):
```
[the remark this answers is not in the document]
Comment (tanemaki) [reply src/anomalib/metrics/threshold/f1_adaptive_threshold.py:78]: Thanks for reviewing this pull request. I agree with your point, so let me modify my code and resend the modified pull request.

File: tests/unit/metrics/test_adaptive_threshold.py (hunk 1 above)
Comment (tanemaki) [reply tests/unit/metrics/test_adaptive_threshold.py:21]: I have just added an explicit check for this case where all images are anomalous 💪 It would be great if you could check if my change is appropriate.

**Comment by djdameln** [comment]
@tanemaki We would like to merge this PR, but your branch is very much out of sync with our latest main branch. I think the easiest way to address this is to create a new branch from main and re-create your changes there (possibly by cherry-picking the commits of your current branch). This approach would prevent having to solve the numerous merge conflicts that will arise when trying to sync your current branch with main.

Please let us know if you're interested in re-submitting your PR. If not, that's totally fine as well, in that case we'll pick up this task on our side 😊

**Comment by samet-akcay** [comment]
it would be good to have @tanemaki as contributor, but if this is not of interest anymore, as @djdameln mentioned, we could pick up the task.

File: tests/unit/metrics/test_adaptive_threshold.py [hunk 3]
```diff
@@ -12,8 +12,10 @@
 @pytest.mark.parametrize(
     ("labels", "preds", "target_threshold"),
     [
-        (torch.tensor([0, 0, 0, 1, 1]), torch.tensor([2.3, 1.6, 2.6, 7.9, 3.3]), 3.3),  # standard case
-        (torch.tensor([1, 0, 0, 0]), torch.tensor([4, 3, 2, 1]), 4),  # 100% recall for all thresholds
+        (torch.Tensor([0, 0, 0, 1, 1]), torch.Tensor([2.3, 1.6, 2.6, 7.9, 3.3]), 3.3),  # standard case
```
Comment (ashwinvaidya17) [inline tests/unit/metrics/test_adaptive_threshold.py]:
```suggestion
        (torch.tensor([0, 0, 0, 1, 1]), torch.tensor([2.3, 1.6, 2.6, 7.9, 3.3]), 3.3),  # standard case
```

File: tests/unit/metrics/test_adaptive_threshold.py [hunk 4]
```diff
@@ -12,8 +12,10 @@
...
+        (torch.Tensor([1, 0, 0, 0]), torch.Tensor([4, 3, 2, 1]), 4),  # 100% recall for all thresholds
```
Comment (ashwinvaidya17) [inline tests/unit/metrics/test_adaptive_threshold.py]:
```suggestion
        (torch.tensor([1, 0, 0, 0]), torch.tensor([4, 3, 2, 1]), 4),  # 100% recall for all thresholds
```

File: tests/unit/metrics/test_adaptive_threshold.py [hunk 5]
```diff
@@ -12,8 +12,10 @@
...
+        (torch.Tensor([1, 1, 1, 1]), torch.Tensor([4, 3, 2, 1]), 1),  # use minimum value when all images are anomalous
```
Comment (ashwinvaidya17) [inline tests/unit/metrics/test_adaptive_threshold.py]:
```suggestion
        (torch.tensor([1, 1, 1, 1]), torch.tensor([4, 3, 2, 1]), 1),  # use minimum value when all images are anomalous
```

File: tests/unit/metrics/test_adaptive_threshold.py [hunk 6]
```diff
@@ -12,8 +12,10 @@
...
+        (torch.Tensor([0, 0, 0, 0]), torch.Tensor([4, 3, 2, 1]), 4),  # use maximum value when all images are normal
```
Comment (ashwinvaidya17) [inline tests/unit/metrics/test_adaptive_threshold.py]:
```suggestion
        (torch.tensor([0, 0, 0, 0]), torch.tensor([4, 3, 2, 1]), 4),  # use maximum value when all images are normal
```

File: src/anomalib/metrics/threshold/f1_adaptive_threshold.py [hunk 7]
```diff
@@ -120,9 +121,26 @@ def compute(self) -> torch.Tensor:
             )
             logging.warning(msg)
 
+            self.value = torch.max(dim_zero_cat(self.preds))
+            return self.value
+
         with handle_mac(self):
             precision, recall, thresholds = super().compute()
```
Comment (rajeshgangireddy) [inline src/anomalib/metrics/threshold/f1_adaptive_threshold.py]: Do we not need the same here?

Comment (ashwinvaidya17) [reply src/anomalib/metrics/threshold/f1_adaptive_threshold.py]: Good catch! Updated

**Review by ashwinvaidya17 — APPROVED** [review]
